### ISSUE TITLE
Discussion for Trilinos as a dependency

### ISSUE THREAD

**Issue by jrood-nrel** [LLNL/spack#4419]
Not sure if discussions go here or outside of github, but anyway, I work on a project that is built on top of Trilinos. I have been putting a lot of effort into having my project use the Trilinos package file in Spack. Although I'm sure it's possible with enough effort and enough variants, I have not been successful with merging my custom Trilinos package file (https://github.com/NaluCFD/NaluSpack/blob/master/spack_config/nalu-trilinos/package.py) into the general one. In my opinion Trilinos is just too big to maneuver and experiment with. It takes over an hour to build the general Trilinos in Spack, and about 25 minutes for my custom Trilinos with only what I need enabled, so it's not easy to test modifications, especially when I need to experiment with modifying dependencies of Trilinos. I don't mind a maximal approach of Trilinos building everything, but I haven't been able to keep track of how many different errors I have encountered when trying to merge my set of options for Trilinos into Spack.

My main question is this: Since my project is very closely tied to Trilinos, what are anyone's thoughts that I just pack Trilinos into my project package file, or is there a higher level way I can select a full set of options I need so that I can simplify this?

Tagging @davydden since I think he has the most experience with Trilinos here.

**Comment by adamjstewart** [comment]
> Not sure if discussions go here or outside of github

Discussions are perfectly welcome here. Some people also use the [Google Group](https://groups.google.com/forum/#!forum/spack) but I personally never check it.

**Comment by jrood-nrel** [comment]
@aprokop Yes I was hoping to discuss this when we meet. Sorry I should have tagged you as well.

**Comment by jrood-nrel** [comment]
@aprokop I'm wondering if I can just implement Trilinos as a resource in Nalu like you had done with DTK in Trilinos, but with the ability to build a package inside a package.

**Comment by aprokop** [comment]
@jrood-nrel It may work. But won't you have to then make Nalu's spack configuration to depend on all of the TPLs that Trilinos requires?

So far, I don't see why we could not merge it with Trilinos proper. I think TPLs are easy, just maybe need to add a bit of package granularity. It would certainly be nice, as personally I'm compiling it in a shared mod, and most of you seem to use static. If you could briefly described the issues you encountered, that would be a start.

**Comment by jrood-nrel** [comment]
@aprokop So below is where I'm at right now and I have decided to switch to my laptop to begin with, so I'm on OSX 10.11. I do not have many previous problems documented because I would just immediately try the next thing. Anyway, the stuff below builds Trilinos for me using the listed dependencies in Nalu, but Nalu complains it can't find `spack/opt/spack/darwin-elcapitan-x86_64/gcc-6.3.0/trilinos-master-onx7offhbl7mv62joxsop4qnuwplwf2z/lib/cmake/Trilinos/../ShyLU/ShyLUConfig.cmake`. If I turn on shared libraries for everything, Trilinos complains that it won't build shared libraries, even when I enable shared libraries all down the dependencies, so I've stuck with static for now.

```
depends_on('yaml-cpp+fpic~shared')    depends_on('trilinos+boost~debug~dtk+tpetra~epetra+exodus+hdf5~hypre~metis+muelu~mumps+pnetcdf~python~shared+stk~suite-sparse+superlu~superlu-dist~xsdkflags@master')
```

```
diff --git a/var/spack/repos/builtin/packages/trilinos/package.py b/var/spack/repos/builtin/packages/trilinos/package.py
index 66dc4f0..ec6e440 100644
--- a/var/spack/repos/builtin/packages/trilinos/package.py
+++ b/var/spack/repos/builtin/packages/trilinos/package.py
@@ -89,6 +89,7 @@ class Trilinos(CMakePackage):
             description='Builds a debug version of the libraries')
     variant('boost',        default=True,  description='Compile with Boost')
     variant('tpetra',       default=True,  description='Compile with Tpetra')
+    variant('epetra',       default=False,  description='Compile with Epetra')
     variant('exodus',       default=False, description='Compile with Exodus from SEACAS')
 
     variant('dtk',          default=False, description='Enable DataTransferKit')
@@ -98,6 +99,9 @@ class Trilinos(CMakePackage):
              placement='DataTransferKit',
              when='+dtk')
     conflicts('+dtk', when='~tpetra')
+    variant('pnetcdf', default=False, description='Compile with PNetCDF')
+    variant('stk', default=False, description='Compile with STK')
+    variant('muelu', default=False, description='Compile with MueLu')
 
     # Everything should be compiled with -fpic
     depends_on('blas')
@@ -112,6 +116,7 @@ class Trilinos(CMakePackage):
     # MPI related dependencies
     depends_on('mpi')
     depends_on('netcdf+mpi')
+    depends_on('netcdf+mpi+parallel-netcdf', when='@master+pnetcdf')
     depends_on('parmetis', when='+metis')
     # Trilinos' Tribits config system is limited which makes it very tricky to
     # link Amesos with static MUMPS, see
@@ -193,7 +198,9 @@ class Trilinos(CMakePackage):
             '-DTrilinos_ENABLE_CXX11:BOOL=ON',
             '-DTPL_ENABLE_Netcdf:BOOL=ON',
             '-DTrilinos_ENABLE_Tpetra:BOOL=%s' % (
-                'ON' if '+tpetra' in spec else 'OFF')
+                'ON' if '+tpetra' in spec else 'OFF'),
+            '-DTrilinos_ENABLE_Epetra:BOOL=%s' % (
+                'ON' if '+epetra' in spec else 'OFF')
         ])
 
         if '.'.join(platform.mac_ver()[0].split('.')[:2]) == '10.12':
@@ -374,6 +381,55 @@ class Trilinos(CMakePackage):
                 '-DTPL_ENABLE_SuperLU:BOOL=OFF',
             ])
 
+        # pnetcdf:
+        if '+pnetcdf' in spec:
+            options.extend([
+               '-DTPL_ENABLE_Pnetcdf:BOOL=ON',
+               '-DPNetCDF_ROOT:PATH=%s' % spec['parallel-netcdf'].prefix
+            ])
+        else:
+            options.extend([
+                '-DTPL_ENABLE_Pnetcdf:BOOL=OFF'
+            ])
+
+        # stk
+        if '+stk' in spec:
+            options.extend([
+                '-DTrilinos_ENABLE_STK:BOOL=ON'
+                '-DTrilinos_ENABLE_STKMesh:BOOL=ON',
+                '-DTrilinos_ENABLE_STKIO:BOOL=ON',
+                '-DTrilinos_ENABLE_STKTransfer:BOOL=ON',
+                '-DTrilinos_ENABLE_STKSearch:BOOL=ON',
+                '-DTrilinos_ENABLE_STKUtil:BOOL=ON',
+                '-DTrilinos_ENABLE_STKTopology:BOOL=ON',
+                '-DTrilinos_ENABLE_STKUnit_tests:BOOL=ON',
+                '-DTrilinos_ENABLE_STKUnit_test_utils:BOOL=ON',
+                '-DTrilinos_ENABLE_STKClassic:BOOL=OFF',
+                '-DTrilinos_ENABLE_STKDoc_tests:BOOL=OFF',
+                '-DTrilinos_ENABLE_Belos:BOOL=ON',
+                '-DTrilinos_ENABLE_Ifpack2:BOOL=ON',
+                '-DTrilinos_ENABLE_Amesos2:BOOL=ON',
+                '-DTrilinos_ENABLE_Zoltan2:BOOL=ON',
+                '-DTrilinos_ENABLE_Ifpack:BOOL=OFF',
+                '-DTrilinos_ENABLE_Amesos:BOOL=OFF',
+                '-DTrilinos_ENABLE_Zoltan:BOOL=ON'
+            ])
+        else:
+            options.extend([
+                '-DTrilinos_ENABLE_STK:BOOL=OFF'
+            ])
+
+        # muelu
+        if '+muelu' in spec:
+            options.extend([
+                '-DTrilinos_ENABLE_MueLu:BOOL=ON',
+            ])
+        else:
+            options.extend([
+                '-DTrilinos_ENABLE_MueLu:BOOL=OFF'
+            ])
+
+
         # python
         if '+python' in spec:
             options.extend([
@@ -405,7 +461,8 @@ class Trilinos(CMakePackage):
         # exodus
         if '+exodus' in spec:
             options.extend([
-                '-DTrilinos_ENABLE_SEACAS:BOOL=ON'
+                '-DTrilinos_ENABLE_SEACAS:BOOL=ON',
+                '-DTPL_ENABLE_X11=OFF'
             ])
         else:
             options.extend([
```

**Comment by davydden** [comment]
@jrood-nrel  @aprokop

i suppose it depends on what type of package are we talking about. If it's similar to DTK (https://github.com/LLNL/spack/pull/4304) and has to be build **while building** trilinos (kind of a plugin), then i suppose adding it inside trilinos as an additional TPL with a dedicated resource is ok. 

Whereas if it is something which is linked against an already installed Trilinos, like `dealii`, then i see no reason why you should not be able to add your package in Spack.

If the problem is that

> Trilinos is just too big to maneuver and experiment with

then I think it's good to add extra variants to make `trilinos` more flexible and lightweight. For example, for `dealii` we only need `Amesos, AztecOO, Epetra, Ifpack, ML, MueLu, Sacado, Teuchos`. So currently it's also a waste of time to build all TPLs in Trilinos.

While adding extra variants for TPLs, one should be careful as not all of them are available for older versions. But this should be easy enough to enforce via
```
conflicts('+muelu', when='@:11.13.99') # MueLu is from 11.14 onwards
```

> Since my project is very closely tied to Trilinos, what are anyone's thoughts that I just pack Trilinos into my project package file

I think this is a wrong way to do it. This is very much against the Spack philosophy and approach that no package should ship their dependencies inside. Otherwise someone gets into big troubles with shared libraries when using your lib `foo` (which happen to contain `trilinos`) as well as some package `bar` which is linked against Spack's `trilinos`.

Bottom line: I am in favour of adding granularity to trlinos' TPL to make it more lightweight and I think it can be done. I never got to look at this closer myself as it was/is not a big enough issue for me.

**Comment by aprokop** [comment]
@jrood-nrel I'll take a look at it at the weekend. Nalu definitely does not need ShyLU right now, so it could be disabled.

**Comment by jrood-nrel** [comment]
Thanks for the pep talk @davyden and @aprokop . I tried a reverse approach of shutting off all packages and only turning on the ones I needed and I think I have had success with this (partly because it cuts the compile time down to 15 minutes). I created pull request #4428 that once finished, should resolve this and then we can close it.

**Comment by jrood-nrel** [comment]
I think we have sufficiently generalized Trilinos to allow Nalu to use it correctly in #4428 .
